Backtracking: Letter Case Permutation

Given a string containing letters and digits, generate all possible strings by changing the case of the letters.

Example

input = "a1b2"

Possible output:

    [
        "a1b2",
        "a1B2",
        "A1b2",
        "A1B2"
    ]

In [54]:
def generateLetterCasePermutations(text, index, current, result):
    if len(current) == len(text):
        return result.append(current)
    if text[index].isalpha():
        current += text[index].lower()
        generateLetterCasePermutations(text, index + 1, current, result)
        current = current[:-1]
        
        current += text[index].upper()
        generateLetterCasePermutations(text, index + 1, current, result)
        current = current[:-1]
    else:
        current += text[index]
        generateLetterCasePermutations(text, index + 1, current, result)
        current = current[:-1]
    return result

print(generateLetterCasePermutations('a1b2', 0, '', []))

['a1b2', 'a1B2', 'A1b2', 'A1B2']


problem: Generate Binary Strings

Given n, generate all binary strings of length n.

For:

n = 2

Output:

    ["00", "01", "10", "11"]

In [58]:
def generateBinaryStrings(n, current, result):
    if len(current) == n:
        return result.append(current)
    
    current += "0"
    generateBinaryStrings(n, current, result)
    current = current[:-1]
    
    current += "1"
    generateBinaryStrings(n, current, result)
    current = current[:-1]
    
    return result

print(generateBinaryStrings(2, "", []))

['00', '01', '10', '11']


Problem: Generate All Subsets

Given an array of distinct numbers, generate all possible subsets.

Example

numbers = [1, 2]

Output:

    [
        [],
        [1],
        [2],
        [1, 2]
    ]

In [60]:
def generateSubsets(numbers, index, current, result):
    if len(numbers) == index:
        return result.append(current)
    generateSubsets(numbers, index + 1, current + [numbers[index]], result)
    generateSubsets(numbers, index + 1, current, result)
    return result

print(generateSubsets([1,2], 0, [], []))

[[1, 2], [1], [2], []]


Problem: Generate All Binary Strings Without Consecutive 1s

Given n, generate all binary strings of length n that do not contain consecutive 1s.

For:

n = 3

Expected:

[
    "000",
    "001",
    "010",
    "100",
    "101"
]

In [11]:
def generateValidBinaryStrings(n, current, result):
    if len(current) == n:
        return result.append(current)
    current += '0'
    generateValidBinaryStrings(n, current, result)
    current = current[:-1]
    if current == '' or current[-1] != '1':
        current += '1'
        generateValidBinaryStrings(n, current, result)
        current = current[:-1]
    return result
    
print(generateValidBinaryStrings(3, '', []))
print(generateValidBinaryStrings(2, '', []))

['000', '001', '010', '100', '101']
['00', '01', '10']


Problem: Generate All Dice Rolls

You are given n, representing the number of times a dice is rolled.

Generate all possible sequences of rolls.

For a simplified dice with only 1 and 2:

n = 2

Expected:

[
    "11",
    "12",
    "21",
    "22"
]

In [19]:
def generateDiceRolls(n, current, result):
    if len(current) == n:
        result.append(current)
        return result
    current += '1'
    generateDiceRolls(n, current, result)
    current = current[:-1]
    
    current += '2'
    generateDiceRolls(n, current, result)
    current = current[:-1]

    return result

print(generateDiceRolls(2, '', []))
print(generateDiceRolls(3, '', []))

['11', '12', '21', '22']
['111', '112', '121', '122', '211', '212', '221', '222']


Problem: Generate Strings of Length n Using A and B, Without Consecutive As

For:

n = 3

Valid output:

[
    "ABA",
    "ABB",
    "BAB",
    "BBA",
    "BBB"
]

Invalid:

"AAA"
"AAB"
"BAA"

In [8]:
def generateValidABStrings(n, current, result):
    if len(current) == n:
        result.append(current)
        return result
    
    current += 'B'
    generateValidABStrings(n, current, result)
    current = current[:-1]
    
    if current == '' or current[-1] != 'A':
        current += 'A'
        generateValidABStrings(n, current, result)
        current = current[:-1]
    
    return result

print(generateValidABStrings(3, '', []))

['BBB', 'BBA', 'BAB', 'ABB', 'ABA']


Problem: Generate All Permutations of "ABC"

Given a string, generate all possible arrangements of its characters.

For:

text = "ABC"

Expected:

[
    "ABC",
    "ACB",
    "BAC",
    "BCA",
    "CAB",
    "CBA"
]

In [24]:
def generateStringPermutations(text, current, result):
    if len(text) == 0:
        result.append(current)
        return result
    for i in range(len(text)):
        current += text[i]
        generateStringPermutations(text[:i] + text[i + 1:], current, result)
        current = current[:-1]
    return result
    
print(generateStringPermutations("ABC", "", []))

['ABC', 'ACB', 'BAC', 'BCA', 'CAB', 'CBA']


Problem: Generate All Strings of `A` and `B` With Exactly `k` A's

Given:

n = 3

k = 2

Generate all strings of length 3 containing exactly two As.

Expected:

["AAB", "ABA", "BAA"]

In [35]:
def generateStringsWithKAs(n, k, current, aCount, result):
    if len(current) == n:
        if aCount == k:
            result.append(current)
        return result
    if aCount < k:
        current += 'A'
        aCount += 1
        generateStringsWithKAs(n, k, current, aCount, result)
        current = current[:-1]
        aCount -= 1
    
    current += 'B'
    generateStringsWithKAs(n, k, current, aCount, result)
    current = current[:-1]
    return result

print(generateStringsWithKAs(3, 2, "", 0, []))

['AAB', 'ABA', 'BAA']


roblem — Generate All Strings of Length n Using A, B, C

For:

n = 2

Expected:

AA
AB
AC
BA
BB
BC
CA
CB
CC

In [42]:
def generateStrings(n, current, result):
    if len(current) == n:
        result.append(current)
        return result
    
    current += 'A'
    # print(current)
    generateStrings(n, current, result)
    current = current[:-1]
    
    current += 'B'
    generateStrings(n, current, result)
    current = current[:-1]
    
    current += 'C'
    generateStrings(n, current, result)
    current = current[:-1]
    return result

print(generateStrings(2, '', []))

['AA', 'AB', 'AC', 'BA', 'BB', 'BC', 'CA', 'CB', 'CC']


Problem — Generate All Strings of Length n Using 0 and 1, With No Restrictions

For:

n = 3

Generate:

000
001
010
011
100
101
110
111

In [45]:
def generateBinaryCombinations(n, current, result):
    if len(current) == n:
        result.append(current)
        return result
    current += '0'
    generateBinaryCombinations(n, current, result)
    current = current[:-1]
    
    current += '1'
    generateBinaryCombinations(n, current, result)
    current = current[:-1]
    
    return result

print(generateBinaryCombinations(3, '', []))

['000', '001', '010', '011', '100', '101', '110', '111']


Problem — Generate All 2-Character Selections

Given:

characters = ["A", "B", "C"]

Generate all possible 2-character arrangements without using the same character twice.

Expected:

AB
AC
BA
BC
CA
CB

In [52]:
def generateSelections(characters, length, current, result):
    if len(current) == length:
        result.append(current)
        return result
    
    for i in range(len(characters)):
        current += characters[i]
        print(i, current)
        generateSelections(characters[:i] + characters[i + 1:], length, current, result)
        current = current[:-1]
    
    return result

print(generateSelections(['A', 'B', 'C'], 2, '', []))

0 A
0 AB
1 AC
1 B
0 BA
1 BC
2 C
0 CA
1 CB
['AB', 'AC', 'BA', 'BC', 'CA', 'CB']
